### PR TITLE
feat: support mapping ObjectId message as mongodb.ObjectId

### PR DESCRIPTION
**Pull request by davearata-snorack**
MongoDB is a popular database for JS / TypeScript projects and the default type for a MongoDB id is an ObjectId. When building protobuf messages from mongo documents it becomes tedious to convert ObjectIds to strings. For example if we have a simple mongo document that has the fields `_id` and `name` like so:
```
{
  _id: ObjectId(),
  name: "foo"
}
```
and a protobuf definition like this:
```
message Document {
  string _id = 1;
  string name = 2;
}
```
You would need to call `_id.toString()` any time you want to send the document as a protobuf message and you would need to call `new mongodb.ObjectId(message._id)` any time you want to convert a message to a mongo document. If it was just one field it would not be so bad. But if you have a more complex document with arrays of ObjectIds or nested objects with ObjectIds you can run into making large functions just for iterating over fields to convert ObjectIds to strings and vice versa.

This change adds an optional flag `useObjectId`. The flag is set to false by default so if the flag is not set then there will be no behavior changes. Similar to the `useDate` flag when the `useObjectId` is set to true any messages with the name `ObjectId` will be mapped to mongodb.ObjectId so that the conversion is handled for the client. This does have an assumption that the `ObjectId` message has one field called `value` that is a `string`. So the protobuf definition would look like this:

```
message ObjectId {
  string value = 1;
}
```
This also requires the client installs the `mongodb` package from npm. There are two tests one where the `ObjectId` message is defined in the same proto file as the message that has an `ObjectId` field. A second test where the `ObjectId` message is defined in an external file.

### CODE REVIEW

**Review by stephenh — COMMENTED** [review]
Cool.

Honestly I'm a little torn to add yet-another-flag into the mix, but I get the value-add and boilerplate reduction you're going after. And this sort of "just hack in your own use cases" is one of the benefits ts-proto tries to provide.

Just brainstorming, it seems like ideally this sort of "bring your own custom types" would be great to somehow be achievable via config settings instead of code changes...

Like basically what you're doing is setting up "for protobuf types that match ...this name..., then use ...this JS type... as the field, with ...this function for encode mapping..., ...this function for decode mapping..., and ...this  function for from json handling...". Etc.

Each of those "...this function..." snippets could probably be ts-poet `imp` strings that would resolve to non-ts-proto-generated files that users would just ship / write themselves...

Dunno @webmaster128 / @boukeversteegh / @aikoven see ^ musing if any of you are particularly excited about running with / would benefit applications you're using ts-proto in. 

But @davearata-snorack , unless you're really jazzed to explore the config-based approach (which would be great! :-)), I'm good with the current approach as well. Maybe we could eventually port it over to the config-based approach if/when that gets implemented.

I had a few minor naming questions, want to remove a few integration tests, but otherwise I'm good approving + shipping this after that.

File: src/types.ts [hunk 1]
```diff
@@ -404,6 +404,11 @@ export function isMapType(ctx: Context, messageDesc: DescriptorProto, field: Fie
   return detectMapType(ctx, messageDesc, field) !== undefined;
 }
 
+export function isObjectId(field: FieldDescriptorProto): boolean {
+  // need to use the includes instead of === because objectid could be imported from an external proto file
+  return field.typeName.includes('.ObjectId');
```
Comment (webmaster128) [inline src/types.ts:409]:
```suggestion
  return field.typeName.endsWith('.ObjectId');
```

?

**Comment by webmaster128** [comment]
I pretty much share @stephenh's concerns here. Maintaining well known types that are shred across the protobuf industry probably makes sense. Adding software specific code will hardly scale.

I think it would be relatively straight forward to create a plugin-system where you map a fully qualified protobuf type name to a plugin file that implements the methods from main.ts.

Is there any official proto definition of an `ObjectId`? Looking at [the type definition itself](https://docs.mongodb.com/manual/reference/bson-types/#objectid), it should be either a 12 bytes `bytes` value or 3 components (timestamp, random, counter) as 3x `bytes`. Writing that as 24 hex byte strings on the wire seems pointless.

**Comment by davearata-snorack** [comment]
@stephenh @webmaster128 Thanks for the feedback! This is my first time working with some of the modules like ts-poet so something like that auto import functionality i wasn't fully aware of.

I hear what you mean when you compare my ObjectId specific implementation vs a generic extensible implementation. I am not sure if I would be the best person to take on that implementation as that seems to be a bit larger of a task and I am not sure I have the availability to take that on, but I do see the value in it.

For now, I will work on the feedback you provided and update the PR.

File: README.markdown [hunk 2]
```diff
@@ -333,6 +338,8 @@ Generated code will be placed in the Gradle build directory.
 
 - With `--ts_proto_opt=useDate=false`, fields of type `google.protobuf.Timestamp` will not be mapped to type `Date` in the generated types. See [Timestamp](#timestamp) for more details.
 
+- With `--ts_proto_opt=useObjectId=true`, fields of a type called ObjectId where the message is constructed to have on field called value that is a string will be mapped to type `mongodb.ObjectId` in the generated types. This will require your project to install the mongodb npm package. See [ObjectId](#objectid) for more details.
```
[the remark this answers is not in the document]
Comment (davearata-snorack) [reply README.markdown:341]: The code relies on the stucture of the objectid protobuf message has that field value that's type string. that's how we convert between a mongodb.ObjectId and a protoObjectId

File: src/main.ts [hunk 3]
```diff
@@ -75,6 +76,10 @@ import { generateGenericServiceDefinition } from './generate-generic-service-def
 export function generateFile(ctx: Context, fileDesc: FileDescriptorProto): [string, Code] {
   const { options, utils } = ctx;
 
+  if (options.useObjectId) {
+    imp('mongodb*mongodb');
```
[the remark this answers is not in the document]
Comment (davearata-snorack) [reply src/main.ts]: removed this line

File: src/main.ts [hunk 4]
```diff
@@ -469,6 +476,46 @@ function makeDeepPartial(options: Options, longs: ReturnType<typeof makeLongUtil
   return { Builtin, DeepPartial, Exact };
 }
 
+function makeObjectIdMethods(options: Options) {
+  const mongodb = imp('mongodb*mongodb');
+
+  const fromObjectId = conditionalOutput(
+    'fromObjectId',
+    code`
+      function fromObjectId(oid: ObjectId): ${mongodb}.ObjectId {
```
[the remark this answers is not in the document]
Comment (davearata-snorack) [reply src/main.ts]: changed this to `fromProtoObjectId`

File: src/main.ts [hunk 5]
```diff
@@ -469,6 +476,46 @@ function makeDeepPartial(options: Options, longs: ReturnType<typeof makeLongUtil
...
+          return ${fromObjectId}(ObjectId.fromJSON(o));
+        }
+      }
+    `
+  );
+
+  const toObjectId = conditionalOutput(
+    'toObjectId',
+    code`
+      function toObjectId(oid: ${mongodb}.ObjectId): ObjectId {
```
[the remark this answers is not in the document]
Comment (davearata-snorack) [reply src/main.ts]: changed this to `toProtoObjectId`

File: src/options.ts [hunk 6]
```diff
@@ -34,6 +34,7 @@ export type Options = {
   forceLong: LongOption;
   useOptionals: boolean | 'none' | 'messages' | 'all'; // boolean is deprecated
   useDate: DateOption;
+  useObjectId: boolean;
```
[the remark this answers is not in the document]
Comment (davearata-snorack) [reply src/options.ts]: changed this to `useMongoObjectId`

File: integration/use-objectid-false/parameters.txt [hunk 7]
```diff
@@ -0,0 +1 @@
+useObjectId=false
```
[the remark this answers is not in the document]
Comment (davearata-snorack) [reply integration/use-objectid-false/parameters.txt]: removed this test

File: README.markdown [hunk 8]
```diff
@@ -671,6 +678,20 @@ The representation of `google.protobuf.Timestamp` is configurable by the `useDat
 | --------------------------- | ---------------------- | ------------------------------------ | ---------------- |
 | `google.protobuf.Timestamp` | `Date`                 | `{ seconds: number, nanos: number }` | `string`         |
 
+## ObjectId
```
[the remark this answers is not in the document]
Comment (davearata-snorack) [reply README.markdown]: removed this part of the doc

File: integration/use-objectid-true/use-objectid-true.proto [hunk 9]
```diff
@@ -0,0 +1,13 @@
+syntax = "proto3";
+
+message ObjectId {
+  string value = 1;
+}
+
+message Todo {
```
[the remark this answers is not in the document]
Comment (davearata-snorack) [reply integration/use-objectid-true/use-objectid-true.proto]: removed this test

File: src/types.ts (hunk 1 above)
Comment (davearata-snorack) [reply src/types.ts]: changed the `includes` to `endsWith`

**Comment by webmaster128** [comment]
I would give this plugin system a shot based on this particular use case. It think it's very helpful for some of the things we are doing as well. In a recent use case, we need to convert decimal types and I'd also love to convert `Timestamps` to [DateWithNanoseconds](https://github.com/cosmos/cosmjs/blob/v0.27.0-rc2/packages/tendermint-rpc/src/dates.ts#L10-L13).

But what I would need for this is 👇. Timestamp can be customized via `google.protobuf.Timestamp`. Is there any fully qualified name for this `message ObjectId` that is widely accepted?

> Is there any official proto definition of an ObjectId?

**Comment by davearata-snorack** [comment]
No I don't believe there is an official proto definition of an ObjectId. I went with a simple
```
message ObjectId {
  string value = 1;
}
```
Because that was how I was converting ObjectId's. My understanding is that it is safe to call `_id.toString()` and `new mongodb.ObjectId(somestring)` and have consistent conversion results. That being said here is the ObjectId typescript definition. Looks like the property `id: Buffer` is the one you are referencing

```
/**
 * A class representation of the BSON ObjectId type.
 * @public
 */
declare class ObjectId {
    _bsontype: 'ObjectId';
    /* Excluded from this release type: index */
    static cacheHexString: boolean;
    /* Excluded from this release type: [kId] */
    /* Excluded from this release type: __id */
    /**
     * Create an ObjectId type
     *
     * @param inputId - Can be a 24 character hex string, 12 byte binary Buffer, or a number.
     */
    constructor(inputId?: string | number | ObjectId | ObjectIdLike | Buffer | Uint8Array);
    /*
    * The ObjectId bytes
    * @readonly
    */
    id: Buffer;
    /*
    * The generation time of this ObjectId instance
    * @deprecated Please use getTimestamp / createFromTime which returns an int32 epoch
    */
    generationTime: number;
    /** Returns the ObjectId id as a 24 character hex string representation */
    toHexString(): string;
    /* Excluded from this release type: getInc */
    /**
     * Generate a 12 byte id buffer used in ObjectId's
     *
     * @param time - pass in a second based timestamp.
     */
    static generate(time?: number): Buffer;
    /**
     * Converts the id into a 24 character hex string for printing
     *
     * @param format - The Buffer toString format parameter.
     */
    toString(format?: string): string;
    /** Converts to its JSON the 24 character hex string representation. */
    toJSON(): string;
    /**
     * Compares the equality of this ObjectId with `otherID`.
     *
     * @param otherId - ObjectId instance to compare against.
     */
    equals(otherId: string | ObjectId | ObjectIdLike): boolean;
    /** Returns the generation date (accurate up to the second) that this ID was generated. */
    getTimestamp(): Date;
    /* Excluded from this release type: createPk */
    /**
     * Creates an ObjectId from a second based number, with the rest of the ObjectId zeroed out. Used for comparisons or sorting the ObjectId.
     *
     * @param time - an integer number representing a number of seconds.
     */
    static createFromTime(time: number): ObjectId;
    /**
     * Creates an ObjectId from a hex string representation of an ObjectId.
     *
     * @param hexString - create a ObjectId from a passed in 24 character hexstring.
     */
    static createFromHexString(hexString: string): ObjectId;
    /**
     * Checks if a value is a valid bson ObjectId
     *
     * @param id - ObjectId instance to validate.
     */
    static isValid(id: string | number | ObjectId | ObjectIdLike | Buffer | Uint8Array): boolean;
    /* Excluded from this release type: toExtendedJSON */
    /* Excluded from this release type: fromExtendedJSON */
    inspect(): string;
}
```

**Comment by boukeversteegh** [comment]
Well, this does kind of point to the need for custom serialization / deserialization logic, but I'm also not in favor of putting application specific transformation logic within ts-proto. A hook would be preferable. Also in this case I would probably go for a generic transformer function that replaces all ObjectIds with strings.

For example (not tested):

```ts
fix(message: any) {
  if (typeof(message) === 'object' && message != null) {
    if (message.objectId) {
      message.objectId = message.objectId.toString();
    }
    Object.values(message).forEach(([key, value]) => fix(value));
  }
}
```

Or another approach, that can be used without modifying ts-proto (this only works for encode/decode):

- Define your messages like this:

```protobuf
message User {
  google.protobuf.Value id= 1;
}
```

- Overwrite the `.wrap` of Value. This is a decorator that detects mongo object ids, and calls toString before passing on.
```ts
    const _wrap = Value.wrap;
    Value.wrap = (value) => {
      return isMongoObjectId(value) ? _wrap(value.toString()) : _wrap(value);
    };
```

Now you can call `User.encode({id: someMongoObjectId})` and it will serialize as string

File: README.markdown (hunk 2 above)
[the remark this answers is not in the document]
Comment (stephenh) [reply README.markdown:340]: Ah yeah, agreed we rely on `.value` but technically the `isObjectId` function is only checking the name of the message type:

https://github.com/stephenh/ts-proto/pull/467/files#diff-c54113cf61ec99691748a3890bfbeb00e10efb3f0a76f03a0fd9ec49072e410aR407

Just the way this is worded, "With ...this setting..., fields called ObjectId where the message ... has  field called value" made me think that `isObjectId` really was going to check both message type name & like `message.fields.length === 1 & message.fields[0].name === value && message.fields[0].type === string`.

**Comment by stephenh** [comment]
Hm, @boukeversteegh I think the `.wrap` hooks would work at runtime, but my assumption is that @davearata-snorack wants the generated `FooMesasge.oid` fields to really be a `monogo.ObjectId` instance, so that passing data in/out of ts-proto/mongo passes type checks.

Also I would assume that the rest of the services within @davearata-snorack 's org already use a protobuf type of `theirPackage.ObjectId`, so switching their `*.proto` files over to something generic like `google.protobuf.Value` is not that doable, at least for this current use case.

Per the plugin system, I agree we should definitely head that way; i.e. I could see something like: `--ts_proto_opt=customTypes=./customTypes.json` where `customTypes.json` looks like something:

```json
{
  "somePackage.ObjectId": {
    type: "MongoDb@mongodb",
    encode: "fromProto@objectIdMapper.ts",
    decode: "toProto@objectIdMapper.ts",
    fromJson: "fromJson@objectIdMapper.ts",
  }
}
```

And `fromProto` / `toProto` / `fromJson` would all be `export`-d top-level functions in the `objectIdMapper.ts` file, that the user would be responsible for creating by hand / separately from the ts-proto output.

But, for now I'm going to merge this PR as-is, because if anything I like to implement things the super-simple / super-hard-coded way first, as @davearata-snorack has done, and then use that to guide making fancier abstractions like the custom types solution (for whoever wants to take it on; I personally probably won't have time soon).

Then, if/when the custom types are implemented, I'd also be fine making a breaking change of removing `useMongoObjectId` and have users like @davearata-snorack just migrate to the new custom types approach.

(If anything, if we don't like it, the existence of `useMongoObjectId` in the codebase will be a good reminder to implement the custom types system. :-D )

**Comment by boukeversteegh** [comment]
> Hm, @boukeversteegh I think the `.wrap` hooks would work at runtime, but my assumption is that @davearata-snorack wants the generated `FooMesasge.oid` fields to really be a `monogo.ObjectId` instance, so that passing data in/out of ts-proto/mongo passes type checks.
> 
> Also I would assume that the rest of the services within @davearata-snorack 's org already use a protobuf type of `theirPackage.ObjectId`, so switching their `*.proto` files over to something generic like `google.protobuf.Value` is not that doable, at least for this current use case.
> 
> Per the plugin system, I agree we should definitely head that way; i.e. I could see something like: `--ts_proto_opt=customTypes=./customTypes.json` where `customTypes.json` looks like something:
> 
> ```json
> {
>   "somePackage.ObjectId": {
>     "type": "MongoDb@mongodb",
>     "encode": "fromProto@objectIdMapper.ts",
>     "decode": "toProto@objectIdMapper.ts",
>     "fromJson": "fromJson@objectIdMapper.ts"
>   }
> }
> ```
> 
> And `fromProto` / `toProto` / `fromJson` would all be `export`-d top-level functions in the `objectIdMapper.ts` file, that the user would be responsible for creating by hand / separately from the ts-proto output.
> 
> But, for now I'm going to merge this PR as-is, because if anything I like to implement things the super-simple / super-hard-coded way first, as @davearata-snorack has done, and then use that to guide making fancier abstractions like the custom types solution (for whoever wants to take it on; I personally probably won't have time soon).
> 
> Then, if/when the custom types are implemented, I'd also be fine making a breaking change of removing `useMongoObjectId` and have users like @davearata-snorack just migrate to the new custom types approach.
> 
> (If anything, if we don't like it, the existence of `useMongoObjectId` in the codebase will be a good reminder to implement the custom types system. :-D )

I like your approach! Premature abstraction often leads to bad design, but to avoid it we have to accept over-specification at first. Thanks for that insight!

**Comment by davearata-snorack** [comment]
@webmaster128 @boukeversteegh @stephenh thanks for the feedback, help and merging the PR! If you decide to go down the path of a more configurable abstraction, I would love to help.

**Pull request merged**